pkg/controller/direct/monitoring/refs.go

```diff
@@ -85,10 +89,76 @@ func normalizeResourceName(ctx context.Context, reader client.Reader, src client
 	return ref, nil
 }
 
+func normalizeMonitoringAlertPolicyRef(ctx context.Context, reader client.Reader, src client.Object, project references.Project, ref *krm.MonitoringAlertPolicyRef) (*krm.MonitoringAlertPolicyRef, error) {
+	if ref == nil {
+		return nil, nil
+	}
+
+	if ref.Name == "" && ref.External == "" {
+		return nil, fmt.Errorf("must specify either name or external on reference")
+	}
+	if ref.Name != "" && ref.External != "" {
+		return nil, fmt.Errorf("cannot specify both name and external on reference")
+	}
+
+	if ref.External != "" {
+		tokens := strings.Split(ref.External, "/")
+		if len(tokens) == 2 && tokens[0] == "alertPolicies" {
+			ref = &krm.MonitoringAlertPolicyRef{
+				External: fmt.Sprintf("projects/%s/alertPolicies/%s", project.ProjectID, tokens[1]),
+			}
+		}
+		if len(tokens) == 4 && tokens[0] == "project" && tokens[2] == "alertPolicies" {
+			ref = &krm.MonitoringAlertPolicyRef{
+				External: fmt.Sprintf("projects/%s/alertPolicies/%s", tokens[1], tokens[3]),
+			}
+		}
+		return nil, fmt.Errorf("format of alertPolicyRef external=%q was not known (use projects/<projectId>/alertPolicies/<alertPolicyId> or alertPolicies/<alertPolicyId>)", ref.External)
+	}
+
+	key := types.NamespacedName{
+		Namespace: ref.Namespace,
+		Name:      ref.Name,
+	}
+	if key.Namespace == "" {
+		key.Namespace = src.GetNamespace()
+	}
+
+	alertPolicy := &unstructured.Unstructured{}
+	alertPolicy.SetGroupVersionKind(schema.GroupVersionKind{
+		Group:   "monitoring.cnrm.cloud.google.com",
+		Version: "v1beta1",
+		Kind:    "MonitoringAlertPolicy",
+	})
+	if err := reader.Get(ctx, key, alertPolicy); err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, fmt.Errorf("referenced MonitoringAlertPolicy %v not found", key)
+		}
+		return nil, fmt.Errorf("error reading referenced MonitoringAlertPolicy %v: %w", key, err)
+	}
+
+	alertPolicyResourceID, err := references.GetResourceID(alertPolicy)
+	if err != nil {
+		return nil, err
+	}
+
+	alertPolicyProjectID, err := references.ResolveProjectIDForObject(ctx, reader, alertPolicy)
+	if err != nil {
+		return nil, err
+	}
+
+	ref = &krm.MonitoringAlertPolicyRef{
```

Comment:
oh, this alertPolicyRef is really complicated (wasn't expect it allows both `projects/<project>/alertPolicies/<id>` and `alertPolicies/<id>`

I have a question about the `AlertPolicy` struct. Since the external has  `projects/<project>/alertPolicies/<id>` I think the resource itself should have a projectRef. But I don't see one https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/e9a7d548ec105d9b36c0c003976b971b8d14027d/pkg/clients/generated/apis/monitoring/v1beta1/monitoringalertpolicy_types.go#L619
Do you have pending PRs or this is a special case?

Reply:
I _think_ that's one of the "older" resources that uses the project-id annotation instead of projectRef.

I also _think_ that's why project-id is listed here: https://cloud.google.com/config-connector/docs/reference/resource-docs/monitoring/monitoringalertpolicy#annotations

Reply:
(And ResolveProjectIDForObject tries projectRef first, then looks at the project-id annotation)

Reply:
That makes sense. Follow up discussion: since we validate and resolve the project from GCP server now, do you think we shall add the `spec.projectRef` for  `MonitoringAlertPolicy`, and only support the previous project-id annotation for backward compatibility reasons? Not a blocker nor a necessary. But I think we have made a lot  of improvements around project for the monitoring dashboard, why not make one more step further.